Add support for large, multi-part config messages

jagerman opened the pull request:
This implements multi-part config messages, which have been stubbed in for a while but still needed an implementation.

Such messages are split after post-compression, post-padding, but before encryption (i.e. each piece gets encrypted separately), and all but the last piece will be a maximum size message (76800 bytes).

The Config object itself keeps track of partially received messages, joining them up as they arrive and processing them once it has a full set (or else expiring them if they are too old).  It also tracks (without storing) recently fully processed multipart messages so that they can be discarded without pointlessly storing them if they happen to arrive again for some reason.

All of this requires some (breaking) external API changes to implement:

- `merge(...)` now returns an ordered_set of config messages that failed to merge rather than a vector; we couldn't keep track of the order anymore, because of the next point.
- `merge(...)` now returns *all* the hashes that make up a multipart set when one gets merged, even if they weren't passed in in the current merge call.  For instance if you have a three-chunk multipart config and merge parts 1 and 2, but not 3, then 1 and 2 will be preserved internally.  When you later provide 3, then all of hash1, hash2, and hash3 get returned to indicate that all of those were processed as a part of this merge.
- confirm_pushed() previously took a vector and implied the order matters (although we never previously returned more than one).  It now takes an unordered_set to be consistent with merge() and current_hashes(), and because the order is irrelevant.
- The C version of confirm_pushed (config_confirm_pushed) previously only accepted a single message hash + size (and would convert this into a one-element vector for confirm_pushed()).  Its API changes quite considerably as it now has to pass an array of hashes and sizes and the size of those arrays.  As with the C++ version, the order of elements here is unimportant.
- The C++ `current_hashes()` previously only ever returned a 0- or 1-length vector; it now returns an unordered_set, and can easily return more than one (i.e. when the current config is a multipart config message).
- Added a new `active_hashes()` method that returns the current hashes *plus* the hashes of any partially loaded multi-part configs for which we are awaiting remaining parts.  When used for "do we want to keep this alive?" this is the more appropriate method to use compared to `current_hashes()`.

Other multi-part related changes (which don't affect the API, but are worth noting):

- multipart messages *do not* support (and must not be) protobuf wrapped, because multipart messages *already* will not be understood by clients not using this new code, and so anything supporting multipart should also have been fixed to not apply extra protobuf wrapping.  We also are returning maximum size message parts, and the protobuf overhead will push it over the storage limit in a way that isn't trivial for libsession-util to determine and compensate for.

Other changes not directly related to multipart, but included as part of this:

- Started converting some use of ustring_view to `span<const unsigned char>` (alias: `ucspan`), adding some helper functions and changing some functions to take spans instead.  For the most part this conversion can happen piecemeal as, luckily, ustring_views are implicitly convertible to `ucspan`'s, and so most places passing a ustring_view (or ustring) will not break.  (The other way, i.e. ucspan to ustring_view, does not work, and so the conversion needs to happen from innermost functions out; this initial change is focussed on some of those inner functions).

- fixed a bug where a duplicate config (multipart or otherwise) that gets merge()d would result in that config's hash ending up in *both* current hashes and old hashes, but it should not have gone into old hashes.

- started converting some string concatenation and std::to_string usage to fmt and/or "..."_format instead now that we have those available (via oxen-logging).

jagerman committed:
Fix compatibility with newer oxenc; add oxen-logging

- Update oxen-encoding to latest dev for new span support in bt code.
- bring in oxen-logging so that we can use fmt; for now this only for
  fmt in the test suite, but we will definitely start using it in the
  future in libsession-util (the libquic PR already does).
- added an overload of the test suite "printable" functions that takes
  format+args.
- replace some `"..."_hex` string concatenations that broke when oxenc
  switched to constexpr _hex no longer return strings.
- Fixed some oxen-related test code that broke where it was touching
  oxenc::detail functions.

jagerman committed:
Add support for large, multi-part config messages

This implements multi-part config messages, which have been stubbed in
for a while but still needed an implementation.

Such messages are split after post-compression, post-padding, but before
encryption (i.e. each piece gets encrypted separately), and all but the
last piece will be a maximum size message (76800 bytes).

The Config object itself keeps track of partially received messages,
joining them up as they arrive and processing them once it has a full
set (or else expiring them if they are too old).  It also tracks
(without storing) recently fully processed multipart messages so that
they can be discarded without pointlessly storing them if they happen to
arrive again for some reason.

All of this requires some (breaking) external API changes to implement:

- `merge(...)` now returns an ordered_set of config messages that failed
  to merge rather than a vector; we couldn't keep track of the order
  anymore, because of the next point.
- `merge(...)` now returns *all* the hashes that make up a multipart set
  when one gets merged, even if they weren't passed in in the current
  merge call.  For instance if you have a three-chunk multipart config
  and merge parts 1 and 2, but not 3, then 1 and 2 will be preserved
  internally.  When you later provide 3, then all of hash1, hash2, and
  hash3 get returned to indicate that all of those were processed as a
  part of this merge.
- confirm_pushed() previously took a vector and implied the order
  matters (although we never previously returned more than one).  It now
  takes an unordered_set to be consistent with merge() and
  current_hashes(), and because the order is irrelevant.
- The C version of confirm_pushed (config_confirm_pushed) previously
  only accepted a single message hash + size (and would convert this
  into a one-element vector for confirm_pushed()).  Its API changes
  quite considerably as it now has to pass an array of hashes and sizes
  and the size of those arrays.  As with the C++ version, the order of
  elements here is unimportant.
- The C++ `current_hashes()` previously only ever returned a 0- or
  1-length vector; it now returns an unordered_set, and can easily
  return more than one (i.e. when the current config is a multipart
  config message).
- Added a new `active_hashes()` method that returns the current hashes
  *plus* the hashes of any partially loaded multi-part configs for which
  we are awaiting remaining parts.  When used for "do we want to keep
  this alive?" this is the more appropriate method to use compared to
  `current_hashes()`.

Other multi-part related changes (which don't affect the API, but are
worth noting):

- multipart messages *do not* support (and must not be) protobuf
  wrapped, because multipart messages *already* will not be understood
  by clients not using this new code, and so anything supporting
  multipart should also have been fixed to not apply extra protobuf
  wrapping.  We also are returning maximum size message parts, and the
  protobuf overhead will push it over the storage limit in a way that
  isn't trivial for libsession-util to determine and compensate for.

Other changes not directly related to multipart, but included as part of
this:

- Started converting some use of ustring_view to `span<const unsigned
  char>` (alias: `ucspan`), adding some helper functions and changing
  some functions to take spans instead.  For the most part this
  conversion can happen piecemeal as, luckily, ustring_views are
  implicitly convertible to `ucspan`'s, and so most places passing a
  ustring_view (or ustring) will not break.  (The other way, i.e. ucspan
  to ustring_view, does not work, and so the conversion needs to happen
  from innermost functions out; this initial change is focussed on some
  of those inner functions).

- fixed a bug where a duplicate config (multipart or otherwise) that
  gets merge()d would result in that config's hash ending up in *both*
  current hashes and old hashes, but it should not have gone into old
  hashes.

- started converting some string concatenation and std::to_string usage
  to fmt and/or "..."_format instead now that we have those available
  (via oxen-logging).

mpretty-cyro commented on include/session/config/base.hpp:
```diff
@@ -188,10 +267,14 @@ class ConfigBase : public ConfigSig {
+    // Invokes the `logger` callback if set, does nothing if there is no logger.  Arguments go
+    // through fmt::format.
+    //
+    // TODO: replace used of this with oxen-logging!
```
I think this has been replaced in the `libQuic` branch (not that that changes anything, but just an FYI)

jagerman replied:
Yeah I thought we had it in *some* branch, just wasn't sure where.  There might be a bit of minor merge conflict around this but hopefully not too bad.

mpretty-cyro commented on include/session/config/base.hpp:
```diff
@@ -1098,6 +1208,29 @@ class ConfigBase : public ConfigSig {
+    ///
+    /// Unlike MULTIPART_MAX_WAIT, such storage does not include the data itself, but merely the
+    /// (final) config hash and timestamp of the recombined parts needed for deduplication.
+    std::chrono::milliseconds MULTIPART_MAX_REMEMBER = 14 * 24h;
```
Is this just an arbitrary duration on the off chance someones device goes offline part way through pushing up a multi-part config and doesn't come back online for a while?

(Just wondering why it's 14 days as opposed to 30 which is the default TTL for config messages)

jagerman replied:
Yes.  14d was a fairly arbitrary choice, and it's really how long we expect to hold onto a set of incomplete chunks that another client uploaded before we give up on them and discard them.

30d would work fine, though; worst case we could just end up holding on to several 75kiB chunks of data for longer.

mpretty-cyro replied:
No worries - I guess it probably doesn't matter _too_ much since if we swap the TTL extension logic to use `active_hashes()` then the parts of the config already on the swarm would stay alive for this expiration + 30 days anyway which should be more than enough time for the original device to come back online and complete the upload

mpretty-cyro commented on include/session/config/base.hpp:
```diff
@@ -917,14 +1001,19 @@ class ConfigBase : public ConfigSig {
+    ///   will be included if the resulting reconstituted multipart message was a valid config; for
+    ///   parts that do not complete a message set, inclusion in the return value is based only on
+    ///   whether the multipart part itself looked valid.
+    std::unordered_set<std::string> merge(
+            const std::vector<std::pair<std::string, ustring>>& configs);
+
+    // Same as above, but takes values as views (because sometimes that is more convenient).
+    std::unordered_set<std::string> merge(
```
I'm sure it'll be fine but we should double check to make sure Desktop & Android can handle getting an `unordered_set` back instead of a `vector`

Bilb replied:
that will be part of the wrapper, and we most likely both will wrap it to a vector/array

jagerman replied:
Yeah there's already a C++ container <-> native container translation that has to happen here.  Depending on how it's written, wrapper code might not even have to change for this one.

mpretty-cyro commented on tests/test_config_contacts.cpp:
```diff
@@ -443,10 +451,428 @@ TEST_CASE("huge contacts compression", "[config][compression][contacts]") {
+    // (which currently isn't compressed) is much larger:
+    CHECK(dump.size() == EXPECT_BIG_DUMP_SIZE);
+
+#if 0
```
What are the `#if 0` for? (just debugging the tests?)

jagerman replied:
Yeah, this is debugging that can go.

jagerman replied:
Also `#if 0` is a proprocessor directive so that unlike `if (0) { ... }` the `...` doesn't have to compile.  (And it wouldn't here, because I deleted the iostream header for std::cerr).

jagerman replied:
Fixed.

Bilb commented on include/session/config.hpp:
```diff
@@ -15,8 +15,20 @@
+// config message into a multipart message.
+inline constexpr int MAX_MESSAGE_SIZE = 76'800;  // 76.8kB = Storage server's limit
+
+// Maximum size of an post-compression outgoing multipart config message (before chopping into
```
```suggestion
// Maximum size of a post-compression outgoing multipart config message (before chopping into
```

jagerman replied:
Fixed

Bilb commented on include/session/config/base.hpp:
```diff
@@ -988,14 +1077,32 @@ class ConfigBase : public ConfigSig {
     ///
-    /// The current config hash(es); this can be empty if the current hash is unknown or the current
-    /// state is not clean (i.e. a push is needed or pending).
+    /// The hashes of the current config state.  This can be empty if the current hashes are not
+    /// known (i.e. a push is still in progress) or the current state is not clean (i.e. a push is
+    /// needed or pending).
+    ///
+    /// See also active_hashes(), which you often want instead of this.
+    ///
+    /// Inputs: None
+    ///
+    /// Outputs:
+    /// - `std::unordered_set<std::string>` -- Returns current config hashes
+    const std::unordered_set<std::string>& current_hashes() const;
```
should we make this protected? I feel like we won't need this anymore for the wrappers but only the active_hashes(), and it would be good to not "use it by mistake".
Or maybe we should rename it, so we are all aware of that change when this PR gets merged

jagerman replied:
I was hoping the type change would be enough for that, but you're right, it might not be.

I've renamed it to `curr_hashes` to induce such breakage and, hopefully, they see the comment about using `active_hashes` instead.

Bilb commented on tests/CMakeLists.txt:
```diff
@@ -30,6 +30,7 @@ target_link_libraries(testAll PRIVATE
     libsession::config
     libsodium::sodium-internal
     Catch2::Catch2WithMain
+    oxen::logging
 )
```
not about that specific file, but I am not too sure where it would be added.
Do we need a test that validates that a dump not multipart is loaded correctly, and saved as multipart?

jagerman replied:
Dumps themselves will are not multipart-aware; the multipart encoding happens when a config needs to be pushed that is too big to fit in one message, and then gets decoded into a single message again when the parts are reassembled just before the data enters the config object.  The only change to dumps is the (new) temporary storage of incomplete partial multiparts that get stored so that we don't have to received all parts in the same call.  For fully reassembled data, the dumped data should look "as if" libsession-util had no size limit and just received and processed a large config data.

Bilb commented:
LGTM. I made a fix for the clang-16 build issues at https://github.com/session-foundation/libsession-util/pull/15

mpretty-cyro commented on src/config/base.cpp:
```diff
@@ -288,30 +563,111 @@ void compress_message(ustring& msg, int level) {
+            remaining = remaining.subspan(chunk_size);
+
+            // We should have filled the buffer exactly:
+            assert(lp.end() == out.data() + out.size());
```
Should this assertion be within the curly braces above? (`lp` isn't defined outside of that scope so I'm getting build errors 😬)

jagerman replied:
Yeah, and it was also broken for the last multipart chunk if the last chunk was relative small.  Fixed.

mpretty-cyro commented on src/config/base.cpp:
```diff
@@ -656,41 +1030,63 @@ LIBSESSION_EXPORT config_push_data* config_push(config_object* conf) {
+        ret->obsolete[i] = reinterpret_cast<char*>(pos);
+        pos += cstr_len;
     }
+    assert(pos - reinterpret_cast<unsigned char*>(ret) == buffer_size);
```
This assertion seems to be getting tripped during the first `config_push(conf)` call iOS makes after creating a new account (for the `UserProfile` config)

The dump for the config is: `64313a21693165313a2439313a64313a23693165313a2664313a2b692d3165313a6e31333a546573744d756c7469436f6e6665313a3c6c6c69306533323aea173b57beca8af18c3519a7bbf69c3e7a05d1c049fa9558341d8ebb48b0c96564656565313a3d646565313a286c65313a296c65313a2a646565`

And the secret key for this empty test account is: `02b272b5eac54c3f67071fdc688fbbbe00000000000000000000000000000000e7efe716cbcf992e57e2c6e94df74b2eddce2074272aca1e5038e5d13b21f715`

Let me know if you need any more info to help debug 🙂

jagerman replied:
Fixed.

mpretty-cyro reviewed (changes requested):
So it looks like this PR doesn't play nicely with the `libQuic` changes due to the updated `oxen-encoding` version (`oxenc::bt_list_consumer btlc(data);` is no longer valid) - it seems we might need to wait until https://github.com/oxen-io/oxen-libquic/pull/152 has been merged?

I worked around this for my testing by modifying `btstream.cpp` to do this instead:
```
auto val = oxenc::const_span<std::byte>{req.data(), req.size()};
oxenc::bt_list_consumer btlc(val);
```
but that resulted in `Cannot create a bt_list_consumer with no data` errors in the longs (and then I wasn't able to continue testing due to the `config_push` assertion but I assume I would have been prevented from testing properly due to this hack)

mpretty-cyro commented on src/config/base.cpp:
```diff
@@ -248,27 +504,45 @@ std::vector<std::string> ConfigBase::_merge(
     } else {
         // the merging affect nothing (if it had seqno would have been incremented), so don't
         // pointlessly replace the inner config object.
         assert(new_conf->unmerged_index() == 0);
```
Just wondering whether this assert should be removed - just thinking that there are situations where a merge can occur with a config which is not going to result in a change

Eg. if you link a new device and, due to network conditions, it wasn't able to retrieve the existing configs then it would push up it's "new user" config, which would then be pulled down on your existing device, merged, and likely result in no changes right?

mpretty-cyro replied:
Actually I just realised that at the moment we consistently merge in the current config because when you push up a new config message we don't update the `lastHash` value (so we don't unintentionally miss a config that was pushed up by another device at the same time)

Which means on the poll after pushing the current config we will always receive the current config back and try to merge it 🙃

mpretty-cyro committed:
Merge branch 'dev' into multipart

# Conflicts:
#	external/CMakeLists.txt
#	external/oxen-encoding
#	include/session/config.hpp
#	include/session/config/base.hpp
#	include/session/config/convo_info_volatile.hpp
#	include/session/config/protos.hpp
#	include/session/hash.hpp
#	include/session/session_encrypt.hpp
#	include/session/types.hpp
#	include/session/util.hpp
#	src/CMakeLists.txt
#	src/config/base.cpp
#	src/config/convo_info_volatile.cpp
#	src/config/encrypt.cpp
#	src/config/internal.cpp
#	src/config/internal.hpp
#	src/config/protos.cpp
#	src/hash.cpp
#	src/session_encrypt.cpp
#	tests/CMakeLists.txt
#	tests/swarm-auth-test.cpp
#	tests/test_blinding.cpp
#	tests/test_bugs.cpp
#	tests/test_config_contacts.cpp
#	tests/test_config_convo_info_volatile.cpp
#	tests/test_config_user_groups.cpp
#	tests/test_config_userprofile.cpp
#	tests/test_configdata.cpp
#	tests/test_group_info.cpp
#	tests/test_group_keys.cpp
#	tests/test_group_members.cpp
#	tests/test_multi_encrypt.cpp
#	tests/test_session_encrypt.cpp
#	tests/utils.hpp

mpretty-cyro committed:
Merge remote-tracking branch 'upstream/dev' into multipart

mpretty-cyro committed:
Fixed C API size calculation

(Jason amended) rewrote the comment describing the struct layout as
well.

jagerman committed:
Add test case trace logger from libquic

This can be very helpful to identify abnormally slow or hanging tests by
trace logging every test as it runs.

jagerman committed:
Speed up huge contacts test with fake session ids

More than half the CPU time in the multi-part config tests is being used
generating 12k proper session IDs (via Ed keygen + conversion to X) for
the test.  This replaces the session id generation with random values
which is sufficient for this test and takes almost no time at all.

jagerman committed:
formatting

jagerman committed:
multiparts: use ordered map for internal storage

We write all held multipart values when dumping with hash as key, but
that won't necessarily be properly sorted with an unordered_map, so make
it sorted instead.

jagerman committed:
More useful test error on network timeout

A timeout would fail all the CHECKs, but would keep going and then throw
a weird json parsing error.  This fixes that by checking the success
condition to a REQUIRE instead of a CHECK so that the json parsing
exception doesn't happen or get reported if the request failed.

mpretty-cyro committed:
Merge remote-tracking branch 'upstream/dev' into multipart

mpretty-cyro committed:
Bumped version to 1.4.0
